### ISSUE TITLE
Updates not installed on restart in 10.12.6

### ISSUE THREAD

**Issue by rodgerramjet26** [homebysix/install-or-defer#4]
I have been testing install-or-defer and it doesn't seem to install updates on restart in 10.12.6 (this was in a VM but don't think that should matter. It worked fine on 10.11.6 on a physical machine.

Looking at the logs (below) it seems to be issues around reloading unloading and loading com.apple.softwareupdated.plist and com.apple.suhelperd.plist.

```
Filtering the log data using "senderImagePath ENDSWITH "logger""
Timestamp                       (process)[PID]    
2017-08-09 15:28:24.586237+0700  localhost logger[1496]: Unloading and removing //Library/LaunchDaemons/com.elliotjordan.install_or_defer.plist...
2017-08-09 15:28:24.758746+0700  localhost logger[1513]: Starting install_or_defer.sh script. Performing validation and error checking...
2017-08-09 15:28:24.866558+0700  localhost logger[1513]: Connection to google.com port 443 [tcp/https] succeeded!
2017-08-09 15:28:24.934685+0700  localhost logger[1513]: Validation and error checking passed. Starting main process...
2017-08-09 15:28:24.934816+0700  localhost logger[1513]: No logo provided. Using App Store icon.
2017-08-09 15:28:24.970505+0700  localhost logger[1513]: Deferral deadline: 2038-01-19 10:14:07
2017-08-09 15:28:24.972464+0700  localhost logger[1513]: Time remaining: 179226h:45m:43s
2017-08-09 15:28:24.976025+0700  localhost logger[1513]: Pre-downloading all available software updates...
2017-08-09 15:43:25.135986+0700  localhost logger[1513]: Software Update Tool
2017-08-09 15:43:25.136745+0700  localhost logger[1513]: 
2017-08-09 15:43:25.137033+0700  localhost logger[1513]: Finding available software
2017-08-09 15:43:25.214216+0700  localhost logger[1513]: Configuring updates to be installed at restart...
2017-08-09 15:43:25.276562+0700  localhost logger[1513]: File Doesn't Exist, Will Create: /var/db/.SoftwareUpdateOptions
2017-08-09 15:43:25.283388+0700  localhost logger[1513]: Reloading com.apple.softwareupdated.plist...
2017-08-09 15:43:27.343689+0700  localhost logger[1513]: /System/Library/LaunchDaemons/com.apple.softwareupdated.plist: service already loaded
2017-08-09 15:43:27.344697+0700  localhost logger[1513]: Reloading com.apple.suhelperd.plist...
2017-08-09 15:43:29.387244+0700  localhost logger[1513]: /System/Library/LaunchDaemons/com.apple.suhelperd.plist: service already loaded
2017-08-09 15:43:29.388373+0700  localhost logger[1513]: Updates configured to install at next restart.
2017-08-09 15:43:29.420746+0700  localhost logger[1513]: Prompting to install updates now or defer...
```

When I the load and unload commands directly on my 10.12.6 VM I got the following errors:
```
sandbox-accounts-Mac:~ sandboxaccount$ sudo launchctl unload /System/Library/LaunchDaemons/com.apple.suhelperd.plist
/System/Library/LaunchDaemons/com.apple.suhelperd.plist: Operation not permitted while System Integrity Protection is engaged
sandbox-accounts-Mac:~ sandboxaccount$ sudo launchctl load /System/Library/LaunchDaemons/com.apple.softwareupdated.plist
/System/Library/LaunchDaemons/com.apple.softwareupdated.plist: service already loaded


sandbox-accounts-Mac:~ sandboxaccount$ sudo launchctl unload /System/Library/LaunchDaemons/com.apple.softwareupdated.plist
/System/Library/LaunchDaemons/com.apple.softwareupdated.plist: Operation not permitted while System Integrity Protection is engaged
sandbox-accounts-Mac:~ sandboxaccount$ sudo launchctl load /System/Library/LaunchDaemons/com.apple.suhelperd.plist
/System/Library/LaunchDaemons/com.apple.suhelperd.plist: service already loaded
```

SIP is enable on both computers so not sure why it is an issue on 10.12.6.

I don't want to disable SIP. Any ideas on if this is a new issue with running install or defer on 10.12.6? Or just an issue for me?

**Comment by mpdonovan** [comment]
I am having the same issue. Has anything been identified as a fix?

**Comment by rodgerramjet26** [comment]
I didn't identify a fix. Instead I used the basic framework/ideas of the script to use softwareupdate to hack together a script to install the updates and continually prompt the user to then restart. It doesn't work anywhere near as well as this install-or-defer does when it works.

**Comment by mpdonovan** [comment]
Sure, I will download a fresh copy and try to do some testing with it tomorrow. Thanks



Mike Donovan
JamfPro Admin /Desktop System Administrator
Killeen Independent School District
104 East Beeline
Harker Heights, Texas 76548




On Jan 10, 2019, at 7:28 PM, Elliot Jordan <notifications@github.com<mailto:notifications@github.com>> wrote:

CAUTION: This email originated from outside of Killeen ISD. DO NOT click links or open attachments unless you recognize the sender and know the content is safe. If you need additional assistance, please contact the Help Desk @ (254) 336-2999.


Hi @rodgerramjet26<https://github.com/rodgerramjet26> and @mpdonovan<https://github.com/mpdonovan> — could you try again now? I've just merged in #7<https://github.com/homebysix/install-or-defer/pull/7>, which makes substantial improvements.

**Comment by mpanighetti** [comment]
New security updates were just released today, so this would be a good time to test them on Sierra.

**Comment by mpanighetti** [comment]
I've confirmed that macOS 10.12.6 is able to apply security updates using the current build of install-or-defer (2.1.3, though it should have been fixed as of 2.0). Closing this out, but pass us a new report if you're still running into issues.
